rfcs/20190411-string-unification.md

```diff
@@ -0,0 +1,302 @@
+is done to convert a
+`TF_STRING` to a `DT_STRING` and vice-versa.  This results in a performance hit
+at external language binding boundaries for string tensors.  Furthermore, the
+current implementation of the C API does not provide setters/getters or other
+ancillary methods for constructing a `TF_STRING`.  As a result, downstream
+language bindings to
+[Java](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/java/src/main/native/tensor_jni.cc?#L270),
+[golang](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/go/tensor.go?#L395),
+etc, modify a raw buffer in order to build
+the index offset list of strings.  A similar conversion is done when TFLite
+strings are passed to C++ kernels.
+
+Our aim with modular TensorFlow is to facilitate the creation of externally
+built and dynamically loaded kernels. With modular TensorFlow, we plan to
+provide a thin header only C++ API that depends on the C API.   If we do not
+update our approach to string tensors for modular TensorFlow, we will incur a
+heavy cost when processing string tensors in kernels due to the constant
+conversion between `DT_STRING` and `TF_STRING` across the C API.  Currently, the
+marshalling of string tensors across TFLite and TFCore incurs a similar cost.
+
+In order to mitigate unnecessary performance degradation, we need to have a
+single definition of a string tensor which is ABI compatible across TFLite, C
+and C++.  Furthermore, the string implementation needs to be ABI compatible
+across various compilers in order to enable modular TensorFlow.
+
+STL containers/strings are not ABI stable, and can vary across compilers,
+compiler versions, and even compiler flags.  To mitigate these issues, we
+propose a lightweight ABI stable replacement for the underlying objects
+representing `DT_STRING`/`TF_STRING`/`kTfLiteString`s.
+
+## Overview
+
+We are proposing two sets of changes in order to (1) unify the definition of
+string tensors across C++/Core, C-API, and TFLite; and (2) to achieve ABI
+stability for string tensors across compilers on a single architecture.
+
+In order to unify the definition of string tensors in C, we propose the addition
+of new methods for creating and ingesting tensors in C.  We are also proposing
+for the original method of creating tensors in the C API to be marked as
+deprecated.  In order to support the transition, we will include a flag in the
+`TF_Tensor` struct to track which API the tensor was created with.  Furthermore,
+we plan to provide accessors and mutators for string tensors in C, in order to
+simplify language bindings, and ease potential future changes to the byte layout
+of string tensors.
+
+For TFLite, we propose a minor change to the layout of a kTfLiteString; the
+prototypes for string creation and string accessors in `strings_util.h` and
+`strings.h` do not need to change.  Existing exported TFLite string tensors will
+need to be updated.
```

Comment:
Just to be clear, we will have to remain backward compatible with the old layout. That is, we'll have to continue supporting execution of models exported with the old layout, as well as the new layout, and be able to distinguish between the two.

Reply:
I updated the doc to address this.